docs/proposals/scheduling/20220530-fine-grained-cpu-orchestration.md

```diff
@@ -0,0 +1,620 @@
+In order to extract the best performance,  optimizations related to CPU isolation,  NUMA-locality are required.  
+
+### Goals
+
+1. Improve the CPU orchestration semantics of Koordinator QoS.
+1. Determine compatible kubelet policies.
+1. Clarified that koordlet should enhance CPU scheduling mechanism.
+1. Provide a set of API such as CPU bind strategies, CPU isolate strategies, NUMA topology alignment strategies, NUMA resources information etc. for applications and cluster administrator to support complex CPU orchestration scenarios.
+1. Provide the CPU orchestration optimization API.
+
+### Non-Goals/Future Work
+
+1. Describe specific design details of koordlet/koord-runtime-proxy.
+1. Describe specific design details of CPU descheduling mechanism.
+
+## Proposal
+
+### Design Overview
+
+![image](/docs/images/cpu-orchestration-seq-uml.svg)
+
+When koordlet starts, koordlet gather the NUMA topology informations from kubelet include NUMA Topology, CPU Topology, kubelet cpu management policy, kubelet allocated CPUs for Guarantee Pods etc. , and update to the NodeResourceTopology CRD.  The latency-sensitive applications are scaling, the new Pod can set Koordinator QoS with LSE/LSR, CPU Bind strategy and CPU Isolate strategy to require koord-scheduler to allocate best-fit CPUs to get the best performance. When koord-scheduler schedling the Pod, koord-scheduler will filter Nodes that satisifed NUMA Topology Alignment strategy, and select the best Node by scoring, allocating the CPUs in Reserve phase, and records the result to Pod annotation when PreBinding. koordlet hooks the kubelet CRI request to replaces the CPUs configuration paramters with the koord-scheduler scheduled result to the runtime such as configurate the cgroup.
+
+### User stories
+
+#### Story 1
+
+Compatible with kubelet's existing CPU management policies. The CPU manager policy `static` allows pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node. If enables the `static` policy, cluster administrator must configure the kubelet reserve some CPUs. There are some options for `static` policy. If the `full-pcpus-only(beta, visible by default)` policy option is specified, the `static` policy will always allocate full physical cores. If the `distribute-cpus-across-numapolicy(alpha, hidden by default)` option is specified, the `static` policy will evenly distribute CPUs across NUMA nodes in cases where more than one NUMA node is required to satisfy the allocation.
+
+
+#### Story 2
+
+Similarly, the semantics of the existing K8s Guarantee Pods in the community should be compatible. The K8s Guarantee Pods like LSE Pods, expect the allocated CPUs are completely exclusive, cannot be shared with best effort workloads to obtain the best performance. But when the load on the node is relatively low, the CPUs allocated by LSR Pods should be shared with best effort workloads to obtain economic benefits. 
```

Comment:
The cpu cores allocated to `guarantee pods` with static policy will not share to the default BestEffort Pods, so it is equivalent to LSE.

Reply:
done